text/0008-partial-schema-validation.md

```diff
@@ -0,0 +1,207 @@
+
+The following declaration for the `User` entity type states that a `User` entity will have exactly one attribute `admin_level`.
+Given this schema, a `User` entity cannot have an `access_level` attribute,
+so the example policy will be rejected because it contains an access to an attribute that cannot exist.
+
+```
+"User": {
+  "shape": {
+    "type": "Record",
+    "attributes": {
+      "admin_level": {"type": "String"}
+    },
+    "additionalAttributes": false,
+  }
+}
+```
+
+## Missing hierarchy
+
+Entity types and actions may be a member of entity types (actions, respectively) beyond those they are explicitly declared to be members of in the schema.
+
+The following policy checks if the principal is a member of the group `Group::"admins"`.
+
+```
+permit(principal in Group::"admins", action, resource);
+```
+
+Suppose that `Group` is not listed in the `memberOfTypes` list for the principal entity type.
+Seeing this, the standard validator would know that the type of the `in` expression is `False` and raise an error stating that this policy can never evaluate to `true`.
+Using partial schema validation, we can specify that the `memberOfTypes` list is incomplete as in the following declaration.
+The partial validator will also assume that any undeclared entity type always has an incomplete `memberOfTypes` list.
+
+```
+"User": { "additionalMemberOfTypes": true, }
+```
+
+With this information, the validator will assume that the principal may be in any group,
+so any `in` expression with a `User` entity as the left operand will have type `Boolean` instead of `False`.
+This will cause the validator to accept the policy so long as the rest of the policy typechecks.
+
+## Schema Format
+
+We extend the schema format to support the above example schema.
+Specifically, we add the following properties to the JSON:
+
+* Any record type, including the record types for entity attributes and action context attributes, may include `"additionalAttributes": true` to specify that values having that record type may have attribute other than those declared.
+* An entity type declaration may include `"additionalMemberOfTypes": true` to specify that entities with that entity type may have ancestor entity types other than those in its  `memberOfTypes` list.
+* An action definition may include `“additionalMemberOf”: true` having the same purpose as `additionalMemberOfTypes` for entity type, but applying to the `memberOf` list.
+
+All properties added to the schema default to `false`,
+so an entity declaration in a schema read by the partial schema validator will be interpreted the same as when read by the standard validator.
+
+
+## Effect on the standard validator
+
+This change does not necessarily require any change to the behavior of the standard validator,
+but the additions to the schema format could be soundly supported.
```

Comment:
Will the implementation make these changes, or is this noting that the changes are possible?